// @rhinestone/module-sdk/_cjs/module/scheduled-orders/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getScheduledOrdersExecutor = void 0;
const viem_1 = require("viem");
const constants_1 = require("./constants");
const uniswap_1 = require("../utils/uniswap");
const getScheduledOrdersExecutor = ({ chainId, executeInterval, numberOfExecutions, startDate, executionData, hook, }) => {
    const swapRouter = uniswap_1.swapRouterAddresses[chainId];
    if (!swapRouter) {
        throw new Error(`Swap router address not found for chainId ${chainId}`);
    }
    return {
        address: constants_1.SCHEDULED_ORDERS_EXECUTOR_ADDRESS,
        module: constants_1.SCHEDULED_ORDERS_EXECUTOR_ADDRESS,
        type: 'executor',
        initData: (0, viem_1.encodePacked)(['address', 'uint48', 'uint16', 'uint48', 'bytes'], [
            swapRouter,
            executeInterval,
            numberOfExecutions,
            startDate,
            executionData,
        ]),
        deInitData: '0x',
        additionalContext: '0x',
        hook,
    };
};
exports.getScheduledOrdersExecutor = getScheduledOrdersExecutor;
