// @rhinestone/module-sdk/_esm/module/multi-factor-validator/installation.js
import { encodeAbiParameters, encodePacked } from 'viem';
import { MULTI_FACTOR_VALIDATOR_ADDRESS } from './constants.js';
export const getMultiFactorValidator = ({ threshold, validators, hook, }) => {
    return {
        address: MULTI_FACTOR_VALIDATOR_ADDRESS,
        module: MULTI_FACTOR_VALIDATOR_ADDRESS,
        initData: encodePacked(['uint8', 'bytes'], [
            threshold,
            encodeAbiParameters([
                {
                    components: [
                        {
                            internalType: 'bytes32',
                            name: 'packedValidatorAndId',
                            type: 'bytes32',
                        },
                        { internalType: 'bytes', name: 'data', type: 'bytes' },
                    ],
                    name: 'validators',
                    type: 'tuple[]',
                },
            ], [validators]),
        ]),
        deInitData: '0x',
        additionalContext: '0x',
        type: 'validator',
        hook,
    };
};
