// @rhinestone/module-sdk/_cjs/module/webauthn-validator/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWebAuthnValidator = void 0;
const viem_1 = require("viem");
const constants_1 = require("./constants");
const getWebAuthnValidator = (webAuthnCredential) => {
    return {
        address: constants_1.WEBAUTHN_VALIDATOR_ADDRESS,
        module: constants_1.WEBAUTHN_VALIDATOR_ADDRESS,
        initData: (0, viem_1.encodeAbiParameters)([
            {
                components: [
                    {
                        name: 'pubKeyX',
                        type: 'uint256',
                    },
                    {
                        name: 'pubKeyY',
                        type: 'uint256',
                    },
                ],
                type: 'tuple',
            },
            {
                type: 'bytes32',
                name: 'authenticatorIdHash',
            },
        ], [
            {
                pubKeyX: BigInt(webAuthnCredential.pubKeyX),
                pubKeyY: BigInt(webAuthnCredential.pubKeyY),
            },
            (0, viem_1.keccak256)((0, viem_1.toHex)(webAuthnCredential.authenticatorId)),
        ]),
        deInitData: '0x',
        hook: webAuthnCredential.hook,
        additionalContext: '0x',
        type: 'validator',
    };
};
exports.getWebAuthnValidator = getWebAuthnValidator;
