// @rhinestone/module-sdk/_esm/module/auto-savings/installation.js
import { encodeAbiParameters } from 'viem';
import { AUTO_SAVINGS_ADDRESS } from './constants.js';
import { swapRouterAddresses } from '../utils/uniswap.js';
export const getAutoSavingsExecutor = ({ chainId, configs, hook, }) => {
    const swapRouter = swapRouterAddresses[chainId];
    if (!swapRouter) {
        throw new Error(`Swap router address not found for chainId ${chainId}`);
    }
    return {
        address: AUTO_SAVINGS_ADDRESS,
        module: AUTO_SAVINGS_ADDRESS,
        initData: encodeAbiParameters([
            { internalType: 'address', name: 'swapRouter', type: 'address' },
            {
                components: [
                    { internalType: 'address', name: 'token', type: 'address' },
                    { internalType: 'uint64', name: 'percentage', type: 'uint64' },
                    { internalType: 'address', name: 'vault', type: 'address' },
                ],
                internalType: 'struct AutoSavings.ConfigWithToken[]',
                name: '_configs',
                type: 'tuple[]',
            },
        ], [swapRouter, configs]),
        deInitData: '0x',
        additionalContext: '0x',
        type: 'executor',
        hook,
    };
};
