// @rhinestone/module-sdk/_esm/module/cold-storage/installation.js
import { encodeAbiParameters, encodePacked, } from 'viem';
import { COLD_STORAGE_HOOK_ADDRESS, COLD_STORAGE_FLASHLOAN_ADDRESS, } from './constants.js';
import { getInstalledModules } from '../../account/index.js';
export const getColdStorageHook = async ({ account, client, waitPeriod, owner, moduleType, hook, }) => {
    const installedModules = await getInstalledModules({ account, client });
    const initData = installedModules.includes(COLD_STORAGE_HOOK_ADDRESS)
        ? '0x'
        : encodePacked(['uint128', 'address'], [BigInt(waitPeriod), owner]);
    return {
        address: COLD_STORAGE_HOOK_ADDRESS,
        module: COLD_STORAGE_HOOK_ADDRESS,
        initData,
        deInitData: '0x',
        additionalContext: '0x',
        type: moduleType,
        hook,
    };
};
export const getAllowedCallbackSenders = ({ addresses, functionSig, callType, selector, hook, }) => {
    return {
        address: COLD_STORAGE_FLASHLOAN_ADDRESS,
        module: COLD_STORAGE_FLASHLOAN_ADDRESS,
        initData: encodeAbiParameters([{ internalType: 'address[]', name: 'addresses', type: 'address[]' }], [addresses]),
        deInitData: '0x',
        functionSig,
        additionalContext: '0x',
        selector,
        hook,
        callType,
        type: 'fallback',
    };
};
